accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -2591,6 +2614,42 @@ Where `<keyword>` is `var` or `final` is treated like so:
     }
     ```
 
+#### Pattern-for-in element
+
+A collection element of the form:
+
+```dart
+for (<keyword> <pattern> in <expression>) <element>
+```
+
+Where `<keyword>` is `var` or `final` is treated like so:
+
+1.  Let `I` be the type of `<expression>`, inferred using the context type
+    schema of `<pattern>`.
+
+2.  Calculate the element type of `I`:
+
+    1.  If `I` implements `Iterable<T>` for some `T` then `E` is `T`.
+
+    2.  Else if `I` is `dynamic` then `E` is `dynamic`.
+
+    3.  Else it is a compile-time error.
+
+3.  Type check `<pattern>` with matched value type `E`.
+
+4.  If there are no compile-time errors, then execution proceeds as the
+    following code, where `id1` and `id2` are fresh identifiers and `append()`
+    is an operation to add an element to the surrounding collection being built:
```

Comment:
We can't use `var` here, because this "rewrite" happens after type inference has already been performed on the expressions.
If we do type analysis again, then we get issues like `<expression>` having two, possibly different, static types. If not, it's undefined what declared type `var id1 = ...` has.
(In general, I prefer to not use desugaring as a specification device if it can be avoidable, but if we do, always give fully typed programs.)

Fx:

Let *E* be the static type of `<element>`.

```dart
  I id1 = <expression>;
  Iterator<T> id2 = id1.iterator;
  while (id2.moveNext()) {
    <keyword> <pattern> = id2.current;
    E id3 = <element>;
    append(id3);
  }
```

Maybe we should just introduce `let` as a specification device, and say that 
```dart
  for (<pattern> in <expression>) <element>
```
desugars (still post inference) to:
```dart
  for (T id in <expression>) let <pattern> = id in <expression>
```

Reply:
> (In general, I prefer to not use desugaring as a specification device if it can be avoidable, but if we do, always give fully typed programs.)

I'm OK with specifying in terms of desugaring as long as we only do that after all type analysis and type errors are done. I think desugaring works pretty well for defining runtime semantics.

I'll go with your first fix, thanks.